subgraphs/bentobox/src/mappings/bentobox.ts

```diff
@@ -10,251 +9,405 @@ import {
   const tokenAddress = event.params.token.toHex()
-  const token = getOrCreateToken(tokenAddress)
 
-  const amount = toDecimal(event.params.amount, token.decimals)
+  const token = getToken(tokenAddress)
 
-  const rebase = getOrCreateRebase(tokenAddress)
-  rebase.elastic = rebase.elastic.minus(amount)
+  const tokenStrategy = getTokenStrategy(tokenAddress)
+
+  const strategyData = getStrategyData(tokenAddress)
+
+  const harvest = getOrCreateHarvest(
+    tokenStrategy.strategy!.concat('-').concat(event.block.number.toString()),
+    tokenStrategy.strategy!,
+    token.id,
+    event
+  )
+
+  const strategyKpi = getStrategyKpi(tokenStrategy.strategy!)
+
+  const rebase = getRebase(tokenAddress)
+  rebase.elastic = rebase.elastic.plus(event.params.amount)
   rebase.save()
 
-  if (!token.strategy) {
-    log.warning(WARNING_MSG_STRATEGY_SET, ['onLogStrategyProfit'])
-    return
-  }
+  const profit = new ProfitOrLoss(tokenStrategy.strategy!.concat('-').concat(strategyKpi.profitOrLossCount.toString()))
+  profit.harvest = harvest.id
+  profit.elastic = rebase.elastic
+  profit.base = rebase.base
+  profit.amount = event.params.amount
+  profit.block = event.block.number
+  profit.timestamp = event.block.timestamp
+  profit.save()
 
-  const strategy = getOrCreateStrategy(token.strategy!, token.id, event.block)
-  strategy.totalProfit = strategy.totalProfit.minus(event.params.amount)
-  strategy.balance = strategy.balance.minus(event.params.amount)
-  strategy.save()
+  harvest.profitOrLoss = profit.id
+  harvest.save()
 
-  createLossStrategyHarvest(token.strategy!, amount, rebase.elastic, event)
-}
+  // If more than one profit or loss for this strategy
+  if (strategyKpi.profitOrLossCount.gt(BigInt.fromU32(0))) {
+    const lastProfitOrLoss = ProfitOrLoss.load(
+      tokenStrategy.strategy!.concat('-').concat(strategyKpi.profitOrLossCount.minus(BigInt.fromU32(1)).toString())
+    )
 
-export function onLogWhiteListMasterContract(event: LogWhiteListMasterContract): void {
-  if (event.params.approved == true) {
-    getOrCreateMasterContract(event.params.masterContract)
-  }
-}
+    const timestampDifference = profit.timestamp.minus(lastProfitOrLoss!.timestamp)
 
-export function onLogSetMasterContractApproval(event: LogSetMasterContractApproval): void {
-  getOrCreateUser(event.params.user)
+    // It's possible that difference can be zero, so ensure it's not before calculating APR
+    if (timestampDifference.gt(BigInt.fromU32(0))) {
+      const multiplier = BigDecimal.fromString('31536000').div(timestampDifference.toBigDecimal())
 
-  const masterContractApproval = getOrCreateMasterContractApproval(event)
-  masterContractApproval.masterContract = event.params.masterContract.toHex()
-  masterContractApproval.approved = event.params.approved
-  masterContractApproval.save()
+      strategyKpi.apr = event.params.amount.toBigDecimal().div(rebase.elastic.toBigDecimal()).times(multiplier)
```

Comment:
@LufyCZ @OlaStenberg this seems incorrect, rebase.elastic has already had the amount added to it at this point, and it's also assuming 100% utilization which is not the case.

Reply:
I think this still suffers from the lack of isolation I mentioned too.

Reply:
After speaking to Lufy he's correct that this accounts for utilization, though this makes it a BentoBox token APR rather than an actual Strategy APR, which is expected as a strategy KPI. I think this ia little confusing, and we might want to store both.